xla/service/gpu/gpu_windowed_einsum_handler.cc

```diff
@@ -163,9 +294,40 @@ absl::StatusOr<bool> GpuWindowedEinsumHandler::Run(
       VLOG(5) << "Processing computation: " << comp->name();
       TF_ASSIGN_OR_RETURN(bool comp_result,
                           HandleAgWindowedEinsumLoop(comp, stream_id));
+      all_ag_loops_.push_back(
+          WindowedEinsumAgLoops(comp->WhileCallInstruction()));
       changed = comp_result;
     }
   }
+  // Now that we have processed all loops, we can check if there are any
+  // allgather-dot pattern that we can optimize. We'd want to transform:
+  //                       input
+  //                       /    |
+  //                      /     |
+  //                     AG    windowed loop
+  //                     /
+  //                    /
+  //                   dot
+  // to:
+  //                       input
+  //                       |
+  //                       |
+  //                     windowed loop
```

Comment:
How does this work? Are you somehow reusing AGs from the loop nest and pipe them through as a result?

Reply:
yes, we gather all slices when we communicate in a ring-fashion, after the whole loop is done, the output buffer will contain the complete activation. This achieves the same result as running AG before the collective matmul transformation.